### PR TITLE
ExecutedBlock cleanup

### PR DESCRIPTION
**Pull request by debris**
changes:
- removed redundant `BlockRef` and `BlockRefMut` structure
- added new structure `Tracing` which represents underlaying data better than `Option<Vec<Vec<FlatTrace>>>`
- removed redundant `bestow_block_reward` function which, I believe, contained bugs. Use machine instead

I next prs ~I will get rid of `BlocksRef` structure and~ I'll try to reduce number of clones we make during block verification

### CODE REVIEW

File: ethcore/src/engines/mod.rs [hunk 1]
```diff
@@ -393,35 +393,3 @@ pub trait EthEngine: Engine<::machine::EthereumMachine> {
...
-	use ethereum_types::{Address, U256};
-
-	/// Give reward and trace.
-	pub fn bestow_block_reward(block: &mut ExecutedBlock, reward: U256, receiver: Address) -> Result<(), Error> {
-		let fields = block.fields_mut();
-
-		// Bestow block reward
-		let res = fields.state.add_balance(&receiver, &reward, CleanupMode::NoEmpty)
-			.map_err(::error::Error::from)
-			.and_then(|_| fields.state.commit());
```
Comment (debris) [inline ethcore/src/engines/mod.rs]: issue 1 -> `bestow_block_reward` is called from `on_close_block` and it does `state.commit()`. It shouldn't. What's more this function is called in a loop, which so it's possible that there were multiple commits made in a single `on_close_block`.

File: ethcore/src/engines/mod.rs [hunk 2]
```diff
@@ -393,35 +393,3 @@ pub trait EthEngine: Engine<::machine::EthereumMachine> {
...
-
-		fields.traces.as_mut().map(move |traces| {
-  			let mut tracer = ExecutiveTracer::default();
-  			tracer.trace_reward(receiver, reward, RewardType::Block);
-  			traces.push(tracer.drain())
-		});
-
-		if let Err(ref e) = res {
-			warn!("Encountered error on bestowing reward: {}", e);
```
Comment (debris) [inline ethcore/src/engines/mod.rs]: issue 2 -> error is silently ignored

File: ethcore/src/block.rs [hunk 3]
```diff
@@ -84,6 +84,38 @@ impl Decodable for Block {
 	}
 }
 
+/// Container for block traces.
+#[derive(Clone)]
+pub enum Tracing {
```
Comment (tomusdrw) [inline ethcore/src/block.rs]: Don't feel like it's a right place to store this structure.

File: ethcore/src/client/client.rs [hunk 4]
```diff
@@ -652,10 +651,7 @@ impl Client {
 
 		// Commit results
 		let receipts = block.receipts().to_owned();
-		let traces = block.traces().clone().unwrap_or_else(Vec::new);
-		let traces: Vec<FlatTransactionTraces> = traces.into_iter()
-			.map(Into::into)
-			.collect();
+		let traces = block.traces().clone().drain();
```
Comment (tomusdrw) [inline ethcore/src/client/client.rs]: Seems like unnecessary clone, since `block` is discarded anyway.

Comment (debris) [reply ethcore/src/client/client.rs]: totally agree!

But for now, we can't omit `clone()`, cause `block` is generic over `IsBlock` trait, and `IsBlock` has only `fn traces(&self)` method.

**Comment by debris** [comment]
> lgtm, although would first like to merge large pending PRs that probably affect similar parts of the code (instead of causing more conflicts)

If you are talking about #7038, it doesn't modify `block.rs` and conflicts should not exist / or be insignificant

File: ethcore/src/client/client.rs (hunk 4 above)
Comment (debris) [reply ethcore/src/client/client.rs]: that's why I wrote, that in next pr 'I'll try to reduce number of clones we make during block verification' ;)

File: ethcore/src/engines/authority_round/mod.rs [hunk 5]
```diff
@@ -991,12 +993,15 @@ impl Engine<EthereumMachine> for AuthorityRound {
 				};
 
 			for empty_step in empty_steps {
-				::engines::common::bestow_block_reward(block, self.block_reward, empty_step.author()?)?;
+				let author = empty_step.author()?;
+				self.machine.add_balance(block, &author, &self.block_reward)?;
+				self.machine.note_rewards(block, &[(author, self.block_reward)], &[])?;
```
Comment (rphmeier) [inline ethcore/src/engines/authority_round/mod.rs]: I think this would be more efficient if we had only one call to `note_rewards`.

**Pull request merged**